alf/environments/mario_wrappers.py

```diff
@@ -0,0 +1,265 @@
+        gym.ActionWrapper.__init__(self, env)
+        # 'B', None, 'SELECT', 'START', 'UP', 'DOWN', 'LEFT', 'RIGHT', 'A'
+        self._num_buttons = len(all_buttons)
+        button_keys = {i for i, b in enumerate(all_buttons) if b in self.BUTTONS}
+        buttons = [(), *zip(button_keys),
+                   *itertools.combinations(button_keys, 2)]
+        # 'UP', 'DOWN', 'LEFT', 'RIGHT'
+        arrows = [(), (4,), (5,), (6,), (7,)]
+        acts = []
+        acts += arrows
+        acts += buttons[1:]
+        acts += [a + b for a in arrows[-2:] for b in buttons[1:]]
+        self._actions = acts
+        self.action_space = gym.spaces.Discrete(len(self._actions))
+
+    def action(self, a):
+        mask = np.zeros(self._num_buttons)
+        for i in self._actions[a]:
+            mask[i] = 1
+        return mask
+
+
+class FrameSkip(gym.Wrapper):
+    """
+    Repeat same action n times and return the last observation
+     and accumulated reward
+    """
+
+    def __init__(self, env, n):
+        gym.Wrapper.__init__(self, env)
+        self.n = n
+
+    def step(self, action):
+        obs = None
+        accumulated_reward = 0
+        done = False
+        info = None
+        for _ in range(self.n):
+            obs, reward, done, info = self.env.step(action)
+            accumulated_reward += reward
+            if done: break
+        return obs, accumulated_reward, done, info
+
+
+class ProcessFrame84(gym.ObservationWrapper):
+    """
+    Resize frame from original resolution to 84x84 or
+    resize to 84x110 and then crop to 84x84
+    """
+
+    def __init__(self, env, crop=True):
+        self.crop = crop
+        super(ProcessFrame84, self).__init__(env)
+        self.observation_space = gym.spaces.Box(
+            low=0, high=255, shape=(84, 84, 1), dtype=np.uint8)
+
+    def observation(self, obs):
+        return ProcessFrame84.process(obs, crop=self.crop)
+
+    @staticmethod
+    def process(frame, crop=True):
+        if frame.size == 210 * 160 * 3:
+            img = np.reshape(frame, [210, 160, 3]).astype(np.float32)
+        elif frame.size == 250 * 160 * 3:
+            img = np.reshape(frame, [250, 160, 3]).astype(np.float32)
+        elif frame.size == 224 * 240 * 3:  # mario resolution
+            img = np.reshape(frame, [224, 240, 3]).astype(np.float32)
+        else:
+            assert False, "Unknown resolution." + str(frame.size)
+        img = img[:, :, 0] * 0.299 + img[:, :, 1] * 0.587 + img[:, :, 2] * 0.114
+        size = (84, 110 if crop else 84)
+        resized_screen = np.array(Image.fromarray(img).resize(
+            size, resample=Image.BILINEAR), dtype=np.uint8)
+        x_t = resized_screen[18:102, :] if crop else resized_screen
+        x_t = np.reshape(x_t, [84, 84, 1])
+        return x_t.astype(np.uint8)
+
+
+#   See https://github.com/openai/baselines/blob/
+#   9b68103b737ac46bc201dfb3121cfa5df2127e53/
+#   baselines/common/atari_wrappers.py#L188-L257
+
+
+class LazyFrames(object):
```

Comment:
remove this class since it will not be used.

Reply:
removed